internal/tracker/downstream_test.go

```diff
@@ -0,0 +1,109 @@
+package tracker
+
+import (
+	"reflect"
+	"sync"
+	"testing"
+)
+
+func TestDownstreamConnsCounts(t *testing.T) {
+	downstream1 := "downstream1"
+	downstream2 := "downstream2"
+
+	tests := []struct {
+		name           string
+		op             func(*DownstreamConns)
+		expectedCounts map[string]uint32
+	}{
+		{
+			name: "record a new downstream connection if under maximum",
+			op: func(tracker *DownstreamConns) {
+				tracker.TryBeginConnection(downstream1, 10)
+			},
+			expectedCounts: map[string]uint32{
+				downstream1: 1,
+			},
+		},
+		{
+			name: "record a connection ending",
+			op: func(tracker *DownstreamConns) {
+				tracker.TryBeginConnection(downstream1, 10)
+				tracker.EndConnection(downstream1)
+			},
+			expectedCounts: map[string]uint32{
+				downstream1: 0,
+			},
+		},
+		{
+			name: "don't record connections which would extend beyond maximum",
+			op: func(tracker *DownstreamConns) {
+				tracker.TryBeginConnection(downstream1, 10)
+				tracker.TryBeginConnection(downstream1, 10)
+				tracker.TryBeginConnection(downstream2, 2)
+				tracker.TryBeginConnection(downstream2, 2)
+
+				// this connection should not be recorded because of the maximums
+				tracker.TryBeginConnection(downstream2, 2)
+				tracker.TryBeginConnection(downstream1, 10)
+
+				tracker.EndConnection(downstream1)
+				tracker.EndConnection(downstream1)
+				tracker.EndConnection(downstream1)
+				tracker.EndConnection(downstream2)
+				tracker.EndConnection(downstream2)
+
+				tracker.TryBeginConnection(downstream2, 2)
+			},
+			expectedCounts: map[string]uint32{
```

Comment:
This test performs many operations. If it fails the expected condition check, debugging will take a while as you won't know which call returned an incorrect value.

Reply:
Yes, I agree its not ideal. I was looking for quick coverage of high level features. If I had more time I would make sure to check the results of tracker.TryBeginConnection(). For now I choose to treat this as a slightly more complex version of `count--` and `count++` and comment the call to TryBeginConnection which should preform anomalously.